include/multipass/ssh/sftp_dir_iterator.h

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (C) 2022 Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#ifndef MULTIPASS_SFTP_DIR_ITERATOR_H
+#define MULTIPASS_SFTP_DIR_ITERATOR_H
+
+#include <libssh/sftp.h>
+
+#include <filesystem>
+#include <functional>
+#include <memory>
+#include <stack>
+#include <vector>
+
+namespace fs = std::filesystem;
+
+namespace multipass
+{
+using SFTPAttributesUPtr = std::unique_ptr<sftp_attributes_struct, std::function<void(sftp_attributes)>>;
+using SFTPDirUPtr = std::unique_ptr<sftp_dir_struct, std::function<int(sftp_dir)>>;
+
+class SFTPDirIterator
+{
+    sftp_session sftp;
+    std::stack<SFTPDirUPtr, std::vector<SFTPDirUPtr>> dirs;
+    SFTPAttributesUPtr previous_attr;
```

Comment:
yeah, i get the confusion. my reasoning was that it holds the previously read attribute, since when returning from `next()`, we either read a new attribute and return the previously read one by swapping it with the newly read one, or we don't read any new attribute and just return the previously read one when we're at the end of a directory. but yeah `next_attribute` makes more sense, since it is always the next attribute that is going to be returned